REF: move copula out of sandbox, and refactor, copula directory

josef-pkt opened the pull request:
same/similar commits as #7292 but moving to distributions.copula directory
(cherry-picked and amended each commit of 7292)

pep8speaks commented:
Hello @josef-pkt! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`statsmodels/distributions/copula/copulas.py`](https://github.com/statsmodels/statsmodels/blob/9270e678152ff7f23b725bd6e379b3d5ccd061c9/statsmodels/distributions/copula/copulas.py):

> [Line 20:1](https://github.com/statsmodels/statsmodels/blob/9270e678152ff7f23b725bd6e379b3d5ccd061c9/statsmodels/distributions/copula/copulas.py#L20): [F401](https://duckduckgo.com/?q=pep8%20F401) 'statsmodels.distributions.copula.depfunc_ev.transform_bilogistic' imported but unused
> [Line 20:1](https://github.com/statsmodels/statsmodels/blob/9270e678152ff7f23b725bd6e379b3d5ccd061c9/statsmodels/distributions/copula/copulas.py#L20): [F401](https://duckduckgo.com/?q=pep8%20F401) 'statsmodels.distributions.copula.depfunc_ev.transform_hr' imported but unused
> [Line 20:1](https://github.com/statsmodels/statsmodels/blob/9270e678152ff7f23b725bd6e379b3d5ccd061c9/statsmodels/distributions/copula/copulas.py#L20): [F401](https://duckduckgo.com/?q=pep8%20F401) 'statsmodels.distributions.copula.depfunc_ev.transform_joe' imported but unused
> [Line 20:1](https://github.com/statsmodels/statsmodels/blob/9270e678152ff7f23b725bd6e379b3d5ccd061c9/statsmodels/distributions/copula/copulas.py#L20): [F401](https://duckduckgo.com/?q=pep8%20F401) 'statsmodels.distributions.copula.depfunc_ev.transform_tawn' imported but unused
> [Line 20:1](https://github.com/statsmodels/statsmodels/blob/9270e678152ff7f23b725bd6e379b3d5ccd061c9/statsmodels/distributions/copula/copulas.py#L20): [F401](https://duckduckgo.com/?q=pep8%20F401) 'statsmodels.distributions.copula.depfunc_ev.transform_tawn2' imported but unused
> [Line 20:1](https://github.com/statsmodels/statsmodels/blob/9270e678152ff7f23b725bd6e379b3d5ccd061c9/statsmodels/distributions/copula/copulas.py#L20): [F401](https://duckduckgo.com/?q=pep8%20F401) 'statsmodels.distributions.copula.depfunc_ev.transform_tev' imported but unused



* In the file [`statsmodels/distributions/copula/transforms.py`](https://github.com/statsmodels/statsmodels/blob/9270e678152ff7f23b725bd6e379b3d5ccd061c9/statsmodels/distributions/copula/transforms.py):

> [Line 12:1](https://github.com/statsmodels/statsmodels/blob/9270e678152ff7f23b725bd6e379b3d5ccd061c9/statsmodels/distributions/copula/transforms.py#L12): [F403](https://duckduckgo.com/?q=pep8%20F403) 'from statsmodels.distributions.copula.depfunc_ev import *' used; unable to detect undefined names
> [Line 12:1](https://github.com/statsmodels/statsmodels/blob/9270e678152ff7f23b725bd6e379b3d5ccd061c9/statsmodels/distributions/copula/transforms.py#L12): [F401](https://duckduckgo.com/?q=pep8%20F401) 'statsmodels.distributions.copula.depfunc_ev.*' imported but unused

josef-pkt commented:
last commit is WIP, in the middle of refactoring when discussion about location came up.
comments about this refactoring of u,v and `args` are in #7292

josef-pkt commented:
no related failures on azure, linting complaints are mainly unused variables.

this has now unit tests for CopulaDistributions cdf against R, using uniform, and for EVtawn also genextreme

next cleanup and then adding missing pieces

For archimedean maybe define logpdf and add logderiv, defined as log(deriv), and same for deriv2 to transformation. Based on browsing some of the code, that should work better than log(pdf)

josef-pkt commented:
need to decide on splitting module

from previous PR

`api.py`
`base.py` top level Copula class and CopulaDistribution class
`copulas.py` mainly api file to collect individual copulas
`_archimedean`, `_extreme_value`, `_other_copulas` as private modules (so we are not committed to location, and extensions)
`transforms` for archimedean  DONE
`ev_depfunction` for ev copula "transforms" (or similar name for module) DONE  name `depfunc_ev`

I'm still not sure
If we have `copulas.py` with copulas, that still leaves `CopulaDistribution` as main user class in `base`

TBC

josef-pkt commented:
my guess right now: 
Elliptical distributions will get frozen correlation or scatter matrix and not part of params, at least at the beginning, and we don't need mean and scale for distribution based copula.
But the parameterization has to match up with ppf of univariate distribution, so we get uniform margins.

josef-pkt commented:
namings for ev copulas

I want to change names for authors to commonly used names in extreme value theory, AsyLogisticCopula or AsymLogisticCopula instead of `tawn` 
`evd` package uses shortcuts e.g. `alog`, `SimCop` package use long names like MVAsyLogistic
`copulas` package uses author names as ev copula option "galambos","gumbel","huslerReiss","tawn","tev" for symmetric versions.
see https://github.com/statsmodels/statsmodels/issues/7258#issuecomment-761079399

We need some alias docs or dict for discoverability, e.g. currently I have `copulanamesbv` dict with some aliases

josef-pkt commented:
I found a Diplomarbeit with derivatives of bivariate extreme value copulas, generic formulas in appendix A
https://mediatum.ub.tum.de/doc/1145695/document.pdf

Also, on page 18 and 19 equ 2.21 - 2.31 has conditional distribution and density for bivariate EV copulas in terms of Pickand's dependence function.
It's nice work that summarizes about ev copulas more than articles or book chapters. Problem is that copula books like Joe 1997 only look at the symmetric cases.

conditional copula/distribution
C2|1(u2|u1) := ∂C(u1, u2) / ∂u1 = C(u1, u2) / u1 * (A(t) − t A'(t))
C1|2(u1|u2) := ∂C(u1, u2) / ∂u2 = C(u1, u2) / u2 * (A(t) + (1 − t)A'(t))

density
c(u1, u2) = C(u1, u2) / u1 u2 * (A(t)**2 + (1 − 2 t) A'(t) A(t) − (1 − t) t (A'(t)**2 - A''(t) / log(u1 u2)
where t = log(u2)/ log(u1 u2)

This needs to wait until I have derivatives of Pickand's dependence functions A

and I ran into Archimax copulas which combine elements of archimedean generator and Pickand's dependence function.
Those copulas can be archimedean, extreme value or neither. (They cannot be both because gumbel is the only one that is both)

In which module would that one go?

josef-pkt commented:
resurrecting another old idea: http://jpktd.blogspot.com/2012/12/visual-inspection-of-random-numbers.html
If we need many random number, we might be better off using interpolation, instead of numerically inverting the cdf separately at each point to get the ppf for sampling.
i.e. we can add an option and memoize the interpolation numbers for 1 set of parameters. Difference to my old experiments is that we simulate in (0, 1) interval instead of real line.

josef-pkt commented:
another idea: plot bivariate marginal cdf in multivariate copula

IIRC, the marginal subset cdf just has ones at all other positions, so we can get a `cdf_marginal_bv` or `cdf_marginal_subset` that we can use in plotting cdf.
I don't remember seeing a simple generic expression for marginal pdf, that might need partial derivatives w.r.t. subsets of variables.
need to verify

josef-pkt commented:
another idea: reusing scipy distribution generic cdf->ppf inversion and rvs
add private module `_conditional` to subclass scipy distributions base class and define only conditional cdf

collection  for 6 cond. cdf is in Joe 1997 p. 147 (only Frank and an "unknown" copula have closed form ppf)

josef-pkt commented:
current limitation 
`args` needs to be a tuple of scalars, i.e. copula args ar not vectorized, even though it doesn't raise exception.
The problem is that broadcasting might work, but requires different vectorization or shape for different methods, ie. 1-dim or 2-dim 1-column arrays.
Current behavior, when it works, is likely not be forwards compatible, when we change to fix this.

Usage, e.g. time varying copula correlation, where copula args are modeled as varying by observation, where we have regression case for copargs and not just marginal args.

josef-pkt commented:
small change in plans: I need to finish bernstein polynomial distributions before I forget again what I read.

Multivariate bernstein distributions based on empirical histogram data do not necessarily have uniform margins, those depend on marginal frequency counts in the data. So they are only approximately a copula.
However, common with copulas is that the support of the distribution is [0,1]^k. i.e. we need to transform the domain for distributions on R or R+.


Bernstein distributions also look good enough, easy enough to implement, that we don't need to keep them as private support tools.

So I prefer to use a `distributions.bernstein` module for both supporting code (e.g. eval function that I had put in distributions.tools) and univariate and multivariate bernstein distributions.

Not clear yet. Bernstein polynomials can also be used for function approximation or `smoother` and other use cases. Those will be outside of `distributions` together with other polynomial bases.
(IIRC, I saw one article that uses BP for estimating, approximating Pickand's dependence functions for EV copulas)

I guess I keep the BP eval functions in tools for now. Those other use cases need extra code for a polynomial basis, derivatives, ...
In my current case, derivatives are just used for pdf, conditional pdf, ...
 I'm not planning to add inference for estimated distribution base on BP  for now. Histogram data, cdf values or cell probabilities/frequencies, are given as known parameters and we just construct the distribution.

josef-pkt commented:
Bug somewhere in bernstein distribution, the variables axis are reversed

from a test case in the works
Frank Copula with two Beta distributions of margins.
If the two beta distributions have the same parameters, i.e. everything exchangable, then results look ok.
When I use different parameters in the two beta marginal distributions, then BersteinDistribution has reversed variables compared to Frank CopulaDistribution.

cdfv are values from copula distribution
cdf_bp are values from bernstein distribution

```
np.max(np.abs(cdf_bp.reshape(grid.k_grid) - cdfv.reshape(grid.k_grid)))
0.29316537173149204

np.max(np.abs(cdf_bp.reshape(grid.k_grid).T - cdfv.reshape(grid.k_grid)))
0.025426007983783472
```

same flipped axis are also visible in the two corresponding cdf contour plots.

Initial unit test for 2-d case uses uniform margins and (exchangeable) Frank copula, so no check for asymmetries across variables.

I might add bivariate version of BernsteinDistribution after all, The code is much simpler than the d-dimensional version because we don't have to handle arbitrary number of dimensions.
(IIRC, I tried out asymmetric cases only using the bivariate bernstein code.)

also: memory issues
eval_bernstein is vectorized in x, but it blows up the memory. I got memory errors or frozen computer with a 501 by 501 grid when evaluating x for all grid points at the same time. 501**4 = 63001502001
Evaluating cdf(x) or pdf(x) is on the user or unit test side, but we might need support for loops or working in batches.

A 51x51 grid works fine, but it is less accurate at turning points, peaks.

josef-pkt commented on statsmodels/distributions/bernstein.py:
```diff
@@ -0,0 +1,83 @@
+        Status: not verified yet except for uniform margins.
+        This uses the smoothed cdf values in the new marginal distribution and
+        not the grid values corresponding to the margin of the multivariate
+        grid. This might change.
```
needs changing

josef-pkt commented:
I'm trying to figure out where I left of before doing the detour to asymmetric kernels

bernstein:

- marginal needs changing to use existing grid values
- I thought I added also a univariate Bernstein distribution, but not in commits
- rvs: only experiments in notebook, not sure about adding a `+1` to params

At the end I was looking at rvs and empirical beta copula in contrast to binned bernstein copula

- empirical beta distribution in `asymmetric` kernel PR is only univariate.
- empirical copula, beta copula by Segers and co-authors and bernstein distribution use product kernel

No experimental code yet for conditional density, or conditional quantile function (ppf) for quasi rvs. I guess I need derivatives of beta/bernstein kernel.
Standard pseudo rvs uses mixture representation using beta distribution.

ev-copulas:

Still need derivatives for pickand's dependence functions, add numerical derivatives generically.
The expressions for analytical first and second derivatives are a bit messy, need mathematica.

Still need gof helper functions for unit testing rvs in copulas and multivariate distributions.

josef-pkt commented:
possible off by one problems: indexing, 0- versus 1-based sums and iteration.
for binned values

Bernstein/Beta case: we need to be careful when evaluating beta distribution, pdf, cdf, rvs that we have the proper bounds, e.g.
  
- rvs we need to stay in the interior  (alpha and beta parameter should not be zero in beta distribution) (*)
- cdf: we include both boundary points of [0, 1]
- pdf: I thought we loose the first point from cdf, but need to check again which points we use in beta.pdf, interior or including bounds.

It's a bit confusing because articles use different indexing in sums, and often switch definitions when going from cdf to pdf.

(*) in Kiriliouk, Segers, Tsukahara 2020  Resampling with Empirical Beta Copulas

observations are indexed i = 1, ... n
beta.rvs has `a + b = n + 1`, and parameters go from a = 1 .. n and b = n .. 1)
where a is rank of data within component and ranks go from 1 .. n.

In Bernstein, we have n bins instead of n observations, but I guess indexing should run the same as in the empirical beta copula/distribution.

If we are not in the copula/ranks setting, then ranks 1...n, will correspond to actual observed values, I guess in analogy to asymmetric beta kernel.
However both beta copula based on ranks and Bernstein distribution use a, equal-spaced grid, (and I guess differs from beta kernel for "empirical" kde)

current code, my beta distribution version in `_eval_berstein_1d` uses parameters that sum to n + 2, where n is number of bins
`poly_base = stats.beta.pdf(xx[..., None], k + 1, n - k + 1) / (n + 1)`
where
`k = np.arange(k_terms).astype(float)`
`n = k_terms - 1.`

(most articles on bernstein are written in `binom` form and not in `beta` form)

josef-pkt commented:
1 million points
bd is bernstein distribution from which points were sample
0.113 seconds to sample 1 million rvs, 1-dim, grid size 101 (100 bins)

![image](https://user-images.githubusercontent.com/440735/111345570-cd515880-8653-11eb-8a74-9f1d26253e85.png)

josef-pkt commented:
finally an example notebook that is not mostly dev code https://gist.github.com/josef-pkt/f6861fdf8c84a5075c43910c452d399f

tupui commented:
> finally an example notebook that is not mostly dev code https://gist.github.com/josef-pkt/f6861fdf8c84a5075c43910c452d399f

Nice examples! Do you have a planing for the integration of these? Let me know if I can help.

bashtage commented:
If the distribution has beta marginal, why is the contour plot not on [0,1]x[0,1]?

bashtage commented:
Oh those are integer indices.

josef-pkt commented:
I didn't specify the x-axis in the contour plots. I have those in another notebook but copy pasted the plot part from an earlier version

josef-pkt commented:
I cleaned up the contourplot and updated the notebook in the gist  (plus some dev code to check get_marginal)

Keeping the same `levels` level sets shows that the peak gets flattened

josef-pkt commented:
I'm stopping with Bernstein for now. BernsteinDistribution has all parts that I want or need for now.
(One part that would be useful to add is automatic batch evaluation of bernstein polynomial to avoid exploding memory usage by accident. I added it to asymmetric kernels, but I don't really want to copy-paste it another 2 to 6 times.)

A bit more work on copula code, then I plan to merge this, so @tupui can merge his code into this.

josef-pkt commented:
derivatives in last commit for pickands dep func are derived in wolfram alpha and edited by hand to get a compact form

- still missing derivative for bilogistic and tev depfuncs 
  I added numerical derivatives. We need numerical derivatives for the fit method, and we better avoid nested num diff.
- tawn or tawn2 still has interchanged a1, a2
- unit tests for ev copula pdf (against R evd package)
- a few pep-8/lint issues

remove old, now likely redundant code: leave for later when we have case specific methods in classes, currently much is generic in ev and archimedean copula classes

josef-pkt commented:
related to pdf of EV copulas: use derivatives of copula function directly if that is available

I haven't looked specifically for this yet. Currently, everything in the EV copulas uses the definition through Pickand's dependence function.
We might get better pdf in some cases if we can derive it directly from the cdf/copula function.

josef-pkt commented:
There is likely a bug in the EV copula generic pdf computation from depfunc.
maybe I copied the formula wrongly from the reference, code looks the same as formula

josef-pkt commented:
I'm not able to match up pdf of tawn2 AsymMixed copula with `evd`, also cdf seems to have flipped arguments if second parameter is nonzero (asymmetric case)

I'm not sure they didn't change some definition or parameterization when computing pdf, cdf.
Their pickand's dep function agrees with mine.
I haven't found another R package that has AsymMixed.

So far, all my evd copula code is still generic with only dep functions as case specific.
The plan is now to unit test generic code, and rely that it works correctly for individual cases, even if it doesn't match evd in all cases.

josef-pkt commented:
checking pdf computed with numerical derivative i.e. 2nd cross derivative in bivariate case
`cev` is a AsymMixed copula with genextreme margins

numdiff pdf agrees with pdf from dep func, (at least at equal u point)
I can use this for consistency testing, independent of other packages

```
cev.cdf([0.5, 0.5], args3), cev.pdf([0.5, 0.5], args3)
(0.3620623461620781, 0.12908851059185103)

p = np.array([0.5, 0.5])
approx_hess(p, cev.cdf, args=(args3,))
array([[-0.12611335,  0.12908851],
       [ 0.12908851, -0.12620085]])
```

josef-pkt commented:
bugfixes and redefinition like flipping asymmetry

Now matches evd for AsymLogistic, AsymNegLogistic, AsymMixed and HR, for cdf and genextreme copula distribution pdf
The two missing AsymBiLogistic and TEV do not have working dep func derivatives yet.

HR and AsymNegLogistic (joe) have inverted dependence parameter, `1/dep` compared to `evd`

The unit test for ev copulas and ev-gev distribution needs to be reorganized, currently a bit messy in the lists for parameterized unit tests.
test_copulas is still not in the copula subfolder.
I guess I will move it after all, because those unit test modules will get larger and need a `results` subdirectory.

josef-pkt commented:
checking briefly book "Statistics of Extremes - Theory and Application" chapter 8 and 9 again
They use  `A(y2 / (y1 + y2))` in the ev copula which is what I initially used, now I have y1 in numerator,  `A(y1 / (y1 + y2))` to match R `evd`

evd has the formula in the notes for `abvevd` function with `A(y1 / (y1 + y2))`

the same will apply to stable dependence function
`l(v1, v2) = (v1 + v2) * A(v2 / (v1 + v2))`  chapter 8, p 268 equ (8,44)
For consistency I will need to change numerator also here.

josef-pkt commented:
all green (1 unrelated test failure with pandas in pre-release testing)

merging,

josef-pkt commented:
@tupui it's merged

tupui commented:
> @tupui it's merged

Great thanks! I will update my PR and let you know then.
